buildSrc/src/main/resources/changelog-schema.json

```diff
@@ -0,0 +1,204 @@
+{
+  "$schema": "http://json-schema.org/draft-07/schema#",
+  "$id": "https://github.com/elastic/elasticsearch/tree/master/docs/changelog",
+  "$ref": "#/definitions/Changelog",
+  "definitions": {
+    "Changelog": {
+      "type": "object",
+      "properties": {
+        "pr": {
+          "type": "integer",
+          "minimum": 1
+        },
+        "issues": {
+          "type": "array",
+          "items": {
+            "type": "integer"
+          }
+        },
+        "area": {
+          "type": "string",
+          "enum": [
+            "Aggregations",
+            "Allocation",
+            "Authorization",
+            "Cluster Coordination",
+            "Distributed",
+            "EQL",
+            "Engine",
+            "Engine changes",
+            "Features/Features",
+            "Features/ILM+SLM",
+            "Features/Indices APIs",
+            "Features/Ingest",
+            "Features/Monitoring",
+            "Geo",
+            "Infra/Core",
+            "Infra/Logging",
+            "Infra/Plugins",
+            "Infra/REST API",
+            "Infra/Scripting",
+            "Infra/Settings",
+            "Machine Learning",
+            "Packaging",
+            "Query Languages",
+            "Ranking",
+            "Rollup",
+            "Search",
+            "Security",
+            "SQL",
+            "Task Management"
+          ]
+        },
+        "type": {
+          "type": "string",
+          "enum": [
```

Comment:
Not from the point-of-view of the docs, no.

Reply:
We have nearly[ 200 pull requests](https://github.com/elastic/elasticsearch/pulls?q=is%3Apr+label%3A%3Eenhancement+label%3A%3Ebreaking) with both the enhancement and breaking labels. How would we deal with them? It doesn't seem right to make this mutually exclusive. Certainly we should support an enhancement PR that included breaking changes as part of this. We should consider these things unique axis I think. Perhaps @elastic/docs might have some insight here.

Reply:
> We should consider these things unique axis I think.

This feels right to me. The current manual process does not treat enhancement, release highlight, or breaking labels as mutually exclusive or even necessarily related.

@debadair has had more recent experience with cleaning up our release docs. I'll defer to her call on what the future state should look like.

Reply:
Strictly speaking, what happens at the moment is that any PR with multiple `>` labels has to be manually categorised by whoever is preparing the release notes. By "release notes", I mean like the [list of changes for 7.12.0](https://www.elastic.co/guide/en/elasticsearch/reference/current/release-notes-7.12.0.html), where each change appears under a single heading, and "Breaking changes" and "Enhancements" are separate sections. That's why I said that from a docs point-of-view, you can only have a single type of change.

We could modify the logic, e.g. if you have a PR with both the `>breaking` and `>enhancement` labels, then you could file the PR under the "Enhancement" section in the release notes but also require the extra section in the yaml file for the breaking information. However, that would be inconsistent with a PR that is only labelled `>breaking`.

Reply:
> We could modify the logic, e.g. if you have a PR with both the >breaking and >enhancement labels, then you could file the PR under the "Enhancement" section in the release notes but also require the extra section in the yaml file for the breaking information.

I think we should support this. That is, for a PR with both labels, it would be listed _twice_ in the changelog, once in each section.

> However, that would be inconsistent with a PR that is only labelled >breaking.

I'm not sure I follow. For a PR that is only `>breaking` it would simply only be listed under breaking changes, no?

Reply:
@jrodewig / @debadair what's your perspective here?

Reply:
> I think we should support this. That is, for a PR with both labels, it would be listed twice in the changelog, once in each section, and with a unique description specific to the type of change.

This makes sense to me. That will also help users who are quickly scanning the RNs.

Reply:
@jrodewig We actually talked this through in the Delivery team sync last week and I didn't update the thread here, apologies. In the meeting, I pointed out that the release documentation already has a place for information about breaking changes or deprecated functionality, separate to the release notes (list of PRs).

Regarding PRs labelled with both e.g. `>enhancement` and `>breaking`, the current process requires the release point person to decide which section each PR should go under, so although we have PRs in GitHub that are labelled this way, to date we've never included them in multiple sections in the release notes. Note that this has resulted in some interesting inconsistencies - #65584 was listed as an enhancement in the list of PRs for release 7.11.0, but had a mention as a breaking change in the migration doc.

In my opinion, this makes sense - an enhancement should be an incremental improvement in some existing functionality. For an enhancement to also be a breaking change seems to me to be an oxymoron.

So with the current code, there's nothing to stop a contributor labelling their PR as `>enhancement`, but still including breaking or deprecation information as well.

@mark-vieira does that agree with your understanding?

Reply:
> In my opinion, this makes sense - an enhancement should be an incremental improvement in some existing functionality. For an enhancement to also be a breaking change seems to me to be an oxymoron.

That makes sense. It seems like we'd want to clarify this with the team and contributors. Adding a check to ensure a PR is not labelled as both an enhancement and breaking change would help enforce that.

Thanks for for explaining!

Reply:
Diving into this late, but if I'm following this correctly, the Migration guide is based on things marked breaking in the changelog. So an enhancement (or even a bug fix) that is a breaking change is not included in either the breaking changes section in the RNs OR the Migration guide? Having inconsistent lists of breaking changes has caused confusion and pain in the past, but having consistent but incomplete lists doesn't feel like the answer.

We don't make breaking changes just for the heck of it, every one is a fix, an enhancement, or a new and better way of doing something. The breaking changes are really more of a filter on the changelog--which of the bug fixes, enhancements, and new features could require a user to update their configuration or code?

Which gets to @geekpete's point about searching and filtering the entries. We want to be able to show users what has changed between the version they are running and the version they want to upgrade to, and then filter that down to just those changes that might require action on their part. And, like we tried to do with the "notable changes", highlight those that are most likely to require action. 

Rather than not allowing enhancement PRs to be tagged as breaking, it feels like we just want to ignore the breaking tag for the purposes of building the release notes, and just use it to generate the content for the migration guide. Which probably means the changelog entries for breaking changes need to an additional piece that provides the "Impact" info that tells users what they need to do.

Reply:
@debadair Here's an annotated example:

```yaml
pr: 69149
issues:
  - 55820
area: Packaging
# In the current scheme, this can only have one value. Breaking changes would always
# be `breaking` or `breaking-java`. 
type: breaking
summary: Remove support for `JAVA_HOME`
versions:
  - v8.0.0

# Supplement sections can also be added. For example:
breaking:
  # Controls whether this entry is included in the `//tag::notable-breaking-changes[]` region
  # in `highlights.asciidoc`. Possible the field name here is misleading.
  notable: false
  title: Remove support for `JAVA_HOME`
  area: Packaging

  # Required
  details: |-
    {es} no longer supports the `JAVA_HOME` environment variable. Note
    that {es} does not treat `JAVA_HOME` being set as a failure, as it is
    perfectly reasonable to have this configured at a system level.

  # Currently required, but I note that this isn't always present in our published
  # release notes where the `details` is short. We could make it optional.
  impact: |-
    If you rely on the `JAVA_HOME` environment variable to configure the
    JDK for {es}, you now need to use the `ES_JAVA_HOME` environment
    variable instead.
```

Where we generate a "starter" YAML for a PR in changes to our `HOMER` bot, we add a `highlight:` section for PRs labelled `release highlight`, a `breaking:` section for PRs labelled `>breaking` or `>breaking-java`, and a `deprecation:` section for PRs labelled `>deprecation`.

We don't yet enforce that e.g. breaking PRs have a `breaking:` section, but we probably should. Conversely, I wasn't planning on prohibiting `breaking:` or `deprecation:` sections where the `type:` is other than `breaking` or `deprecation`, leaving that up to a PR author's discretion.

Provided all the PRs in a release provide these YAML files and the contents are good, then we can generate the release documents under `docs/reference/release-notes/` and `docs/reference/migration` (or at least a first pass that is 95% of the way there).

The restriction in this PR on having a single `type` value comes from our existing release notes process. Right now, any PR with multiple type labels (e.g. `>breaking` and `>enhancement`) has to be placed under a single heading by the release point person. They could pick `=== Breaking changes` or they could pick `=== Enhancements`. I wanted to push that decision to the PR author, not the release point person.

We could allow multiple values for `type:`, but to me that implies a change in the list of PRs page, and I was aiming to automate the process without significantly changing the output.

Note that there's nothing here to stop someone using multiple type labels on a PR, they'll just need to pick one in the changelog YAML.